### ISSUE TITLE
13B Model Out of Memory with Single Node 8 A100 GPUs

### ISSUE THREAD

**Issue by benathi** [EleutherAI/gpt-neox#409]
Hi!

Thanks for contribution making this repo available :)

I tried to train the 13B model with micro batch size 1, model parallelism degree 8, but unable to get it to work. (always get OOM) The library advertises being able to scale up to 100B. What is required for this? I also tried deepspeed stage 3 with offload without using pipeline parallelism but that doesn't seem to work either. Please let me know what I'm missing. thanks!

**Comment by StellaAthena** [comment]
Can you post the exact config file you are using?

**Comment by EricHallahan** [comment]
Can you also provide details of your hardware?

**Comment by benathi** [comment]
I’m using an AWS p4 node with 8 of A100 GPUs :)

**Comment by benathi** [comment]
I adapted the config provided 


# GPT-2 pretraining setup
{
   # parallelism settings ( you will want to change these based on your cluster setup, ideally scheduling pipeline stages
   # across the node boundaries )
   "pipe-parallel-size": 1,
   "model-parallel-size": 8,

   # model settings
   "num-layers": 40,
   "hidden-size": 5120,
   "num-attention-heads": 40,
   "seq-length": 2048,
   "max-position-embeddings": 2048,
   "norm": "layernorm",
   "pos-emb": "rotary",
   "no-weight-tying": true,

   # these should provide some speedup but takes a while to build, set to true if desired
   "scaled-upper-triang-masked-softmax-fusion": false,
   "bias-gelu-fusion": false,

   # optimizer settings
   "optimizer": {
     "type": "Adam",
     "params": {
       "lr": 0.0001,
       "betas": [0.9, 0.999],
       "eps": 1.0e-8,
     }
   },
   "zero_optimization": {
    "stage": 1,
    "allgather_partitions": True,
    "allgather_bucket_size": 500000000,
    "overlap_comm": True,
    "reduce_scatter": True,
    "reduce_bucket_size": 500000000,
    "contiguous_gradients": True,
    "cpu_offload": False
  },

   # batch / data settings
   "train_micro_batch_size_per_gpu": 1,
   "data-impl": "mmap",
   "split": "949,50,1",

   # activation checkpointing
   "checkpoint-activations": true,
   "checkpoint-num-layers": 1,
   "partition-activations": true,
   "synchronize-each-layer": true,

   # regularization
   "gradient_clipping": 1.0,
   "weight-decay": 0,
   "hidden-dropout": 0,
   "attention-dropout": 0,

   # precision settings
   "fp16": { 
     "fp16": true,
     "enabled": true,
     "loss_scale": 0,
     "loss_scale_window": 1000,
     "hysteresis": 2,
     "min_loss_scale": 1
   },

   # misc. training settings
   "train-iters": 320000,
   "lr-decay-iters": 320000,
   "distributed-backend": "nccl",
   "lr-decay-style": "cosine",
   "warmup": 0.01,
   "save-interval": 10000,
   "eval-interval": 1000,
   "eval-iters": 10,

   # logging
   "log-interval": 100,
   "steps_per_print": 10,
   "keep-last-n-checkpoints": 4,
   "wall_clock_breakdown": true,
}

**Comment by sweinbach** [comment]
Also note that the above config has a train_micro_batch_size_per_gpu of 1. On 8 gpus that results in a data parallel of 1 (8 / pipe-parallel-size / model-parallel-size) and hence a batch size of 1. I suggest to find a good combination of micro batch size and gradient-accumulation steps to get a decent batch size. See [here](https://github.com/EleutherAI/gpt-neox/blob/e1c82509f23d5316c75c441aa9f49c2142d40f3f/megatron/neox_arguments/arguments.py#L461) for calculation.


Having said that, a 13B will take a long time to train on 8 GPUs only.

**Comment by benathi** [comment]
Thank you! I’ll try that and will let you know. Does the original config 13B.yml also run on a single node with 8GPUs / what hardware setup was it tested on? Thanks again for a fast reply :)

**Comment by sweinbach** [comment]
Have not tested the 13B on a single node with 8 A100s. It is somewhat tricky to balance, too. Reason are the relatively large embedding and lm-head layers that take a lot of memory.

**Comment by benathi** [comment]
I tried setting scaled-upper-triang-masked-softmax-fusion and bias-gelu-fusion to True but it doesn't seem to work, with the provided script for 13B or my modified script with lowering batch size to 1. 

What hardware set up is the 13B config provided tested on btw? If it's not that many nodes I can try to replicate. Is the memory reduction from mostly Zero 1 as it splits opt across nodes? 

Thanks :)

**Comment by sweinbach** [comment]
I don't know the smallest hardware people have tried it on. On 8 GPUs I would estimate a training time of ~2 years though (16 GPUs ~ 1year). Seems not feasible.

**Comment by seeEssex** [comment]
Hi!

Would like to ask if you managed to get it work eventually, Thanks.

**Comment by StellaAthena** [comment]
@seeEssex Why do you want to do this? It would take years to train the model even if it were to be made to fit.

**Comment by seeEssex** [comment]
@StellaAthena I was trying to fit the model for finetuning, as opposed to training the whole thing.

**Comment by StellaAthena** [comment]
@seeEssex there does not currently exist a public 13B model to finetune. The only model we have released so far that is larger than [GPT-J](https://huggingface.co/EleutherAI/gpt-j-6B) is a 20B parameter model. I do know someone who is finetuning it, and can inquire about their hardware and performance.

**Comment by jennyzzt** [comment]
What is the hardware and speed for the person fine-tuning the 20B parameter model above?

I am interested in using GPT-NeoX-20B for fine-tuning. Would 2 servers of 8 A100s GPU be sufficient?
